# About

This repository contains the scripts and IPython notebooks that can be used to reproduce the results by Ditzler and Rosen (2014). The work focuses on things to consider when examining feature subset selection when applied to abundance data in the microbiome. Our goal in this work is to examine the role, assumptions, and inferences that feature subset selection can provide the field of microbial ecology and comparative metagenomics. Specifically we examine feature subset selection algorithms using embedded and filter approaches to infer taxa importance on data collected from the human gut microbiome.  

The code in this repository is licensed under the [GNU GPL v3](http://www.gnu.org/copyleft/gpl.html). 

# Dependencies 

* [scikit-learn](http://scikit-learn.org/stable/)
* [pandas](http://pandas.pydata.org/)
* [fizzy](https://github.com/EESI/Fizzy)
* [qiime](http://qiime.org/) - just for `supervised_learning.py` 
* (*optional*) [IPython](ipython.org) - only if you're interested in running the IPython notebooks.

# Citation Policy 

If you the scripts scripts in this work, and you plan to publish a manuscript based on these codes, please add the following citation: 

* [Gregory Ditzler](http://gregoryditzler.com) and [Gail Rosen](http://www.ece.drexel.edu/gailr/index.html), ``Feature Subset Selection for Inferring Relative Importance of Taxonomy,'' to appear *ACM International Workshop on Big Data in Life Sciences*, 2014.

